bpf/utils.h

```diff
@@ -337,4 +338,129 @@ static inline bool check_and_do_flow_filtering(flow_id *id, u16 flags) {
     return false;
 }
 
+static inline void core_fill_in_l2(struct sk_buff *skb, flow_id *id, u16 *family) {
+    struct ethhdr eth;
+
+    __builtin_memset(&eth, 0, sizeof(eth));
+
+    u8 *skb_head = BPF_CORE_READ(skb, head);
+    u16 skb_mac_header = BPF_CORE_READ(skb, mac_header);
+
+    bpf_probe_read(&eth, sizeof(eth), (struct ethhdr *)(skb_head + skb_mac_header));
+    __builtin_memcpy(id->dst_mac, eth.h_dest, ETH_ALEN);
+    __builtin_memcpy(id->src_mac, eth.h_source, ETH_ALEN);
+    id->eth_protocol = bpf_ntohs(eth.h_proto);
+    if (id->eth_protocol == ETH_P_IP) {
+        *family = AF_INET;
+    } else if (id->eth_protocol == ETH_P_IPV6) {
+        *family = AF_INET6;
+    }
+}
+
+static inline void core_fill_in_l3(struct sk_buff *skb, flow_id *id, u16 family, u8 *protocol,
+                                   u8 *dscp) {
+    u16 skb_network_header = BPF_CORE_READ(skb, network_header);
+    u8 *skb_head = BPF_CORE_READ(skb, head);
+
+    switch (family) {
+    case AF_INET: {
+        struct iphdr ip;
+        __builtin_memset(&ip, 0, sizeof(ip));
+        bpf_probe_read(&ip, sizeof(ip), (struct iphdr *)(skb_head + skb_network_header));
+        __builtin_memcpy(id->src_ip, ip4in6, sizeof(ip4in6));
+        __builtin_memcpy(id->dst_ip, ip4in6, sizeof(ip4in6));
+        __builtin_memcpy(id->src_ip + sizeof(ip4in6), &ip.saddr, sizeof(ip.saddr));
+        __builtin_memcpy(id->dst_ip + sizeof(ip4in6), &ip.daddr, sizeof(ip.daddr));
+        *dscp = ipv4_get_dscp(&ip);
+        *protocol = ip.protocol;
+        break;
+    }
+    case AF_INET6: {
+        struct ipv6hdr ip;
+        __builtin_memset(&ip, 0, sizeof(ip));
+        bpf_probe_read(&ip, sizeof(ip), (struct ipv6hdr *)(skb_head + skb_network_header));
+        __builtin_memcpy(id->src_ip, ip.saddr.in6_u.u6_addr8, IP_MAX_LEN);
+        __builtin_memcpy(id->dst_ip, ip.daddr.in6_u.u6_addr8, IP_MAX_LEN);
+        *dscp = ipv6_get_dscp(&ip);
+        *protocol = ip.nexthdr;
+        break;
+    }
+    default:
+        return;
+    }
+}
+
+static inline void core_fill_in_tcp(struct sk_buff *skb, flow_id *id, u16 *flags) {
+    u16 skb_transport_header = BPF_CORE_READ(skb, transport_header);
+    u8 *skb_head = BPF_CORE_READ(skb, head);
+    struct tcphdr tcp;
+    u16 sport, dport;
+
+    __builtin_memset(&tcp, 0, sizeof(tcp));
+
+    bpf_probe_read(&tcp, sizeof(tcp), (struct tcphdr *)(skb_head + skb_transport_header));
```

Comment:
I just noticed that in utils.h [here](https://github.com/netobserv/netobserv-ebpf-agent/blob/main/bpf/utils.h#L47-L84) , we do a similar extraction of the headers but always with a protective block to ensure we're not trying to read more than `skb->data_end`. Here and below we don't do that, but should we?

Reply:
I don't think data_end is populated for the rtt and psample hook CORE function should be safe otherwise the verifier would screamed at us